kore/src/Kore/Builtin/Int.hs

```diff
@@ -325,3 +331,32 @@ builtinFunctions =
     log2 n
         | n > 0 = Just (smallInteger (integerLog2# n))
         | otherwise = Nothing
+
+evalEq :: Builtin.Function
+evalEq resultSort arguments@[_intLeft, _intRight] =
+    concrete <|> symbolicReflexivity
+  where
+    mkCeilUnlessDefined termLike
+      | TermLike.isFunctionalPattern termLike = Condition.topOf resultSort
+      | otherwise =
+        Condition.fromPredicate (makeCeilPredicate resultSort termLike)
+    returnPattern = return . flip Pattern.andCondition conditions
+    conditions = foldMap mkCeilUnlessDefined arguments
+
+    concrete = do
+        _intLeft <- expectBuiltinInt eqKey _intLeft
+        _intRight <- expectBuiltinInt eqKey _intRight
+        _intLeft == _intRight
+            & Bool.asPattern resultSort
+            & return
+
+    symbolicReflexivity =
```

Comment:
Even though we add side conditions to make sure the inputs are defined, we still need to make sure that the inputs are function-like patterns:

```suggestion
    symbolicReflexivity =
        Control.Monad.guard (TermLike.isFunctionPattern _intLeft)
        -- Do not need to check _intRight because we only return a result
        -- when _intLeft and _intRight are equal.
```

Reply:
Actually, `mkCeilUnlessDefined` uses `isFunctionalPattern`. I forgot to change its name :sweat_smile: . Should I make it check only if the inputs are defined and use the guard you suggested for function-like check?